### PR TITLE
KAFKA-6647 KafkaStreams.cleanUp creates .lock file in directory it tries to clean

### PR DESCRIPTION
**Pull request by tedyu**
Specify StandardOpenOption#DELETE_ON_CLOSE when creating the FileChannel.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by gbloggs** [comment]
Whilst this may be a nice to have I remain unconvinced that this will fix the issue. 
Looking at clean() in the StateDirectory, it catches any IOExceptions thrown from Utils.delete(). Utils.delete  in postVisitDirectory will throw java.nio.file.DirectoryNotEmptyException so a new StreamsException(e) will be thrown

```java
public synchronized void clean() {
        try {
            cleanRemovedTasks(0, true);
        } catch (final Exception e) {
            // this is already logged within cleanRemovedTasks
            throw new StreamsException(e);
        }
        try {
            Utils.delete(globalStateDir().getAbsoluteFile());
        } catch (final IOException e) {
            log.error("{} Failed to delete global state directory due to an unexpected exception", logPrefix(), e);
            throw new StreamsException(e);
        }
    }
```

**Comment by mjsax** [comment]
\cc @bbejeck @vvcephei

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StateDirectory.java [hunk 1]
```diff
@@ -138,7 +138,7 @@ synchronized boolean lock(final TaskId taskId) throws IOException {
         }
 
         try {
-            lockFile = new File(directoryForTask(taskId), LOCK_FILE_NAME);
+            lockFile = new File(stateDir, taskId + LOCK_FILE_NAME);
```
Comment (bbejeck) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StateDirectory.java:145]: why remove the `directoryForTask` method?

**Comment by bbejeck** [comment]
Left one minor comment.  Plus do we need to add a new test?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StateDirectory.java (hunk 1 above)
Comment (tedyu) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StateDirectory.java:145]: The location of the lock file is lifted one level up.

**Comment by tedyu** [comment]
This seems related:
https://stackoverflow.com/questions/2479222/problem-with-java-file-locking-mechanism-filelock-etc

**Review by guozhangwang — COMMENTED** [review]
I left a comment on https://issues.apache.org/jira/browse/KAFKA-6647 ticket itself.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StateDirectory.java (hunk 1 above)
Comment (mjsax) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StateDirectory.java:145]: After discussion on #4713 I think this idea should actually work. Nit: Can we rename the lock to `LOCK_FILE_NAME + " -" + taskId` though.

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/StateDirectoryTest.java [hunk 2]
```diff
@@ -88,27 +89,6 @@ public void shouldCreateTaskStateDirectory() {
         assertTrue(taskDirectory.isDirectory());
     }
 
-    @Test
-    public void shouldLockTaskStateDirectory() throws IOException {
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/StateDirectoryTest.java:92]: Why did you remove this test?

**Comment by tedyu** [comment]
Guozhang has different opinion on the approach.
Waiting for discussion to conclude before taking further action.

**Comment by gartho** [comment]
Any update on this issue?

**Comment by guozhangwang** [comment]
@gartho We are still pondering what is the best fix for this issue across different file systems. What's your use case and is this a blocker issue for your app?

**Comment by gartho** [comment]
I’m coming across this issue during integration level testing. I’ve decided to resolve it for now by clearing out the state store folder before I run my tests so it’s not blocking me at present.

**Comment by knieschraube** [comment]
Any updates so far?

**Comment by knieschraube** [comment]
Any updates?

**Comment by guozhangwang** [comment]
@gartho  @philippkamps  Sorry for being late on this PR. @tedyu I'd suggest follow the option 2) I've proposed in the JIRA ticket, i.e.

```
Just another note, for locking purposes, when the apps shutdown cleanly, it is OK to have the lock file left in the state directory since threads are excluding each other not via the ownership of the file but via locking the file handle. So for your case, if you indeed want to clean up the whole directory upon shutting down, then I think there is still a valid point to close all the file channels upon shutting down. For which we can consider:

1) either use StandardOpenOption.DELETE_ON_CLOSE as you did in the PR.
2) or add a new function in the state directory class that closes all the managed file channels upon KafkaStreams.close(); which may be safer than 1) since StandardOpenOption.DELETE_ON_CLOSE is best-effort.
```

WDYT?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StateDirectory.java (hunk 1 above)
Comment (tedyu) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StateDirectory.java:145]: I can rename the lock if people think we can continue with this solution.

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/StateDirectoryTest.java (hunk 2 above)
Comment (tedyu) [reply streams/src/test/java/org/apache/kafka/streams/processor/internals/StateDirectoryTest.java:92]: directory.lock(taskId) is at different level than channel.tryLock() , rendering the test useless.

**Comment by pgrabarsky** [comment]
+1

**Comment by mjsax** [comment]
@tedyu What is the status of this PR?

**Comment by tedyu** [comment]
From @guozhangwang 's comment on the JIRA, I thought he acknowledged that lifting the lock up one dir would solve the problem for Windows environment.

**Comment by mjsax** [comment]
I just lost the overview :) Last comment from @guozhangwang (https://github.com/apache/kafka/pull/4702#issuecomment-407633443) seems to indicate that he wants some updates to the PR. Or is this already contained and the PR is ready for reviewing?

Maybe you can rebase the PR to resolve the merge conflict first? Thanks a lot!

**Comment by tedyu** [comment]
Can we discuss how to achieve compatibility with existing locking structure first ?

Thanks

**Comment by mjsax** [comment]
Sure. However, I am not exactly sure what you mean? Can you elaborate?

**Comment by tedyu** [comment]
I meant that, after the move of the lock file, is there any compatibility issue we need to consider.

**Comment by tedyu** [comment]
I didn't find the local repo for the original PR.

Need to open a new one if the approach in current PR is confirmed.

**Comment by mjsax** [comment]
@tedyu Sorry for late reply.

> I meant that, after the move of the lock file, is there any compatibility issue we need to consider.

I don't think so. If an app is shut down, it should free all locks. On startup, it can just use the new locking strategy. Or do I miss anything? It would only be an issue, if multiple thread (or different versions) would operate in parallel -- but this case should be be possible on a single machine and thus we should be fine?

**Comment by tedyu** [comment]
bq. but this case should be be possible on a single machine

I guess you meant 'should not be'

**Comment by tedyu** [comment]
Created #5650

**Comment by guozhangwang** [comment]
@tedyu @mjsax My concern about compatibility is that if we have multiple instances of Streams running on the same machine, say A and B, where A shuts down and upgrades to a new version, it will then try to create the lock file in a different directory (one layer up) while B is still holding locks on the old directory, is it possible that both instances think they now own the task's directory and move on to modify the state (e.g. one still owning the task while the other moves in and "cleanup" the state)?

**Comment by mjsax** [comment]
If you run two `KafkaStreams` instanced on one machine, you need to configure it with two different `state.dir` -- thus, both should be isolated and the case you describe cannot happen. Or am I wrong with the assumption about different `state.dir`s ? IIRC, we consider using the same `state.dir` a configuration error.

**Comment by bbejeck** [comment]
I just started up a small example app that merely created the `state.dir`.  I ran two instances of the application. 

In one case I used the same app-id, and both instances started up fine.  It makes sense as the different instances will end up writing to different task-id directories.

In the other case, I changed the app-id of the second instance and a new directory `state.dir/app-id/<task-directories>`.

So I think we need more testing to cover the scenario @guozhangwang outlined above.  But it wouldn't hurt to follow @mjsax suggestion to configure with different `state.dir` entries, but it would be easy to forget to do so.

**Comment by guozhangwang** [comment]
Today we do not actually enforce different `state.dir` configs for different instances: imagine if Streams is specified via K8s, where the config values are wired in and are the same. K8s may allocate containers on the same machine.

Thoughts @vvcephei @bbejeck @tedyu @dguy ?

**Comment by tedyu** [comment]
I think for K8s deployment, the state directory should be customized for each streaming app.

**Comment by bhavesh3184** [comment]
@tedyu What is the status of this PR?

**Comment by tedyu** [comment]
See the above comment from Guozhang.
It seems we should consider the case where state.dir is shared between streaming apps

**Comment by mjsax** [comment]
Can we move the discussion to the new PR? Thanks.
